calendar_resource/models/calendar_event.py

```diff
@@ -0,0 +1,268 @@
+                if categs:
+                    raise ValidationError(
+                        _(
+                            "The resource, '%s', is not allowed in the "
+                            "following event types: \n%s",
+                        )
+                        % (
+                            resource.name,
+                            ', '.join([
+                                categ.name for categ in categs
+                            ]),
+                        )
+                    )
+
+    @api.multi
+    @api.constrains('resource_ids', 'start', 'stop')
+    def _check_resource_ids_leaves(self):
+
+        for record in self:
+
+            if record._event_in_past():
+                continue
+
+            for resource in record.resource_ids:
+
+                if not resource.calendar_id:
+                    continue
+
+                if not resource.calendar_id.leave_ids:
+                    continue
+
+                conflict_leaves = resource.calendar_id.leave_ids.filtered(
+                    lambda s: s.date_from < record.stop and
+                    s.date_to > record.start
+                )
+
+                if not conflict_leaves:
+                    continue
+
+                datetimes = [(c.date_from, c.date_to) for c in conflict_leaves]
+                raise ValidationError(
+                    _(
+                        "The resource, '%s', is on leave during "
+                        "the following times which are conflicting with "
+                        "this event.\n%s",
+                    )
+                    % (
+                        resource.name,
+                        self._format_datetime_intervals_to_str(datetimes),
+                    )
+                )
+
+    @api.multi
+    def _get_event_date_list(self):
+        """ Builds a list of datetimes of the days of the event
+
+        Each datetime in the list is the beginning of a
+        separate day.
+
+        """
+        self.ensure_one()
+        start_date = fields.Date.from_string(self.start)
+        stop_datetime = fields.Datetime.from_string(self.stop)
+
+        if stop_datetime.time() == time(0, 0):
+            stop_datetime -= timedelta(days=1)
+
+        return list(
+            rrule(DAILY, dtstart=start_date, until=stop_datetime.date())
+        )
+
+    @api.multi
+    @api.constrains('resource_ids', 'start', 'stop')
+    def _check_resource_ids_working_times(self):
+
+        for record in self:
+
+            if record._event_in_past():
+                continue
+
+            event_start = fields.Datetime.from_string(record.start)
+            event_stop = fields.Datetime.from_string(record.stop)
+
+            event_days = record._get_event_date_list()
+
+            for resource in record.resource_ids.filtered(
+                    lambda s: s.calendar_id):
+
+                available_intervals = []
+                conflict_intervals = []
+
+                for day in event_days:
```

Comment:
To much nested for, you should consider use searchs

Reply:
One is to iterate through the records @api.multi